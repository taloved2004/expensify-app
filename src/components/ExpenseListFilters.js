import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import {setTextFilter} from '../actions/filters';
import {sortByAmount, sortByDate, setStartDate, setEndDate,setMaxAmount,setMinAmount,setNoteFilter} from '../actions/filters';

export class ExpenseListFilters extends React.Component{
    state = {
        calendarFocused: null
    };
    
    onDatesChange= ({startDate, endDate}) =>{
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onFocusChange = (calendarFocused) =>{
        this.setState(() => ({ calendarFocused }));
    }
    onSortChange = (e)=>{
        e.target.value==='date'?
        this.props.sortByDate() 
        :
        this.props.sortByAmount()
    }
    onTextChange = (e)=>{
        this.props.setTextFilter(e.target.value);
    };
    onNoteChange = (e)=>{
        this.props.setNoteFilter(e.target.value);
    };
    onMinAmountChange = (e)=>{
        e.target.value === ''?
        this.props.setMinAmount(0)
        :
        this.props.setMinAmount(e.target.value);
    };
    onMaxAmountChange = (e)=>{
        e.target.value === ''?
        this.props.setMaxAmount(0)
        :
        this.props.setMaxAmount(e.target.value);
    };
    render(){
        return (
            <div className="content-container">
                <div >
                    <h2 className ="input-group__title">Search Option</h2>
                </div>
                <div className="input-group-1">
                    <div className="input-group__item">
                        <input 
                            type="text" 
                            placeholder="Search description" 
                            value={this.props.filters.text} 
                            onChange={this.onTextChange}
                            className="text-input"
                        />
                    </div>
                    <div className="input-group__item">
                        <input 
                            type="text" 
                            placeholder="Search note" 
                            value={this.props.filters.note} 
                            onChange={this.onNoteChange}
                            className="text-input"
                        />
                    </div>
                   <div className="input-group__item">
                       <input 
                            type="number" 
                            placeholder="Min amount" 
                            onChange={this.onMinAmountChange}
                            className="text-input"
                        />
                    </div>

                    <div className="input-group__item">
                        <input 
                            type="number" 
                            placeholder="Max amount" 
                            onChange={this.onMaxAmountChange}
                            className="text-input"
                        />
                    </div>
                </div>
                <div className="input-group-2">
                    <div className="input-group__item">
                        <select 
                            className="select"
                            value={this.props.filters.sortBy} 
                            onChange={this.onSortChange}
                         >
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker 
                            startDate = {this.props.filters.startDate}
                            endDate = {this.props.filters.endDate}
                            onDatesChange = {this.onDatesChange}
                            focusedInput = {this.state.calendarFocused}
                            onFocusChange = {this.onFocusChange}
                            numberOfMonths = {1}
                            isOutsideRange = {()=>false}
                            showClearDates = {true}
                    />
                    </div>
                </div>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) =>({
    setTextFilter : (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setMaxAmount: (maxMount) => dispatch(setMaxAmount(maxMount)),
    setMinAmount: (minMount) => dispatch(setMinAmount(minMount)),
    setNoteFilter: (note) => dispatch(setNoteFilter(note))
    
});

const mapStateToProps = (state) =>{
    return {
        filters: state.filters
    }
};

export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilters);