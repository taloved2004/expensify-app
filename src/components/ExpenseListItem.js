import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
const Item = ({id, description, amount, createdAt,note}) =>(
        <Link className="list-item" to={`/edit/${id}`} >
            <div>
                <h3 className="list-item__title" >description: {description}</h3>
                <span className="list-item__subtitle">{moment(createdAt).format('MMMM Do, YYYY')}</span>
                {note&&<p className="list-item__subdata">notes: {note}</p>}
            </div>
            <h3 className="list-item__data">
                {numeral(amount/100).format('$0,0.00')}
            </h3>
        </Link>
);

export default Item;