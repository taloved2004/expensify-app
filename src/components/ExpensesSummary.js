import React from 'react';
import { connect } from 'react-redux';
import selectedExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';
import {Link} from 'react-router-dom';

export const ExpensesSummary = ({expenseCount, expensesTotal, hiddenExpenses}) =>{
    const expenseWord = expenseCount===1? 'expense': 'expenses';
    const formattedExpensesTotal = numeral(expensesTotal/100).format('$0,0.00');     
    const msgForHiddenExpenses = hiddenExpenses === 1? 'expense' : (hiddenExpenses>1 ? 'expenses': undefined);
    return (
        <div className="page-header">
            <div className="content-container">
                  <h1 className="page-header__title">Viewing <span>{expenseCount}</span> {expenseWord} totaling <span>{formattedExpensesTotal}</span></h1>
                  {hiddenExpenses === 0 ?
                    <p><b><i>Showing all expenses.</i></b></p>
                    :
                    <p className="page-header__title"><i>Hiding <span>{hiddenExpenses}</span> {msgForHiddenExpenses}.</i></p>}
                  <div className="page-header__actions">
                        <Link className="button" to="/create">Add Expense</Link>
                  </div>
            </div>
        </div>
    );
};

const mapStateToProps = (state) =>{
    const visibleExpenses = selectedExpenses(state.expenses, state.filters);
    return {
        hiddenExpenses: state.expenses.length - visibleExpenses.length,
        expenseCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses)
    };
};
export default connect(mapStateToProps)(ExpensesSummary);