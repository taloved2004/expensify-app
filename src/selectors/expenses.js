//Get visible expenses
import moment from 'moment';

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate,note, minAmount, maxAmount } ) =>{
    return expenses.filter((expense)=>{

        const createdAtMoment = moment(expense.createdAt);
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
        const startDateMatch = startDate? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch =  endDate? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const noteMatch = note? expense.note.toLowerCase().includes(note.toLowerCase()) : true;
        const minAmountMatch = minAmount !== 0? (minAmount*100)<=expense.amount : true;
        const maxAmountMatch =  maxAmount !== 0? expense.amount<=(maxAmount*100) : true;
        return textMatch && startDateMatch && endDateMatch&&noteMatch&&minAmountMatch&&maxAmountMatch;
    }).sort((a,b)=>{
        if(sortBy==='date'){
            return a.createdAt<b.createdAt? 1: -1;
        }
        if(sortBy==='amount'){
            return a.amount<b.amount? 1 : -1;
        }
    });
};

export default getVisibleExpenses;