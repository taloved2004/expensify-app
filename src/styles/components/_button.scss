.button{
    background: $light-blue;
    font-size: $font-size-large;
    font-weight: 300;
    padding: $s-size;
    color: white;
    border: none;
    line-height:1;
    display:inline-block;
    text-decoration: none;
}

.button--cancel{
    background-color: gray;
    margin-left: $s-size;
}

.button--link{
    background:none;
    color: white;
    border: none;
}

.button--secondery{
    
    background:darken($gray,15%);
    margin-bottom: $l-size;
}