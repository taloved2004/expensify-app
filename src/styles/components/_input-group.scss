.input-group-1{
    display: flex;
    flex-direction: column;
    margin-bottom: $m-size;
    @media (min-width:$desktop-breakpoint) {
        display: flex;
        margin-bottom: $l-size;
        flex-direction: row;
    }
}

.input-group-2{
    @extend .input-group-1;
    @media (max-width:$desktop-breakpoint)  {
            margin-top: 0;
       }
}

.input-group__title{
    text-decoration: underline;
    font-weight: 700;
    //padding: $s-size;
}

.input-group__item{
    margin-bottom: $s-size;
    @media (min-width:$desktop-breakpoint) {
        margin:0 $s-size 0 0;
    }
}

