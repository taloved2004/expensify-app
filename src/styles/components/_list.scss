.list-header{
    background: $off-white;
    border: 1px solid darken($off-white, 7%);
    color:$dark-gray-2;
    display: flex;
    font-weight: 600;
    justify-content: space-between;
    padding: $s-size $m-size
}

.list-body{
    margin-bottom: $m-size;
    @media (min-width:$desktop-breakpoint){
        margin-bottom: $l-size;        
    }
}

.list-item{
    border:1px solid darken($off-white, 7%);
    border-top:none;
    color: $dark-gray;
    display: flex;
    flex-direction: column;
    padding:$s-size;
    text-decoration: none;
    transition: background .3s ease;
    &:hover{
        background:$off-white;
    }
    @media (min-width:$desktop-breakpoint){
        align-items: center;
        flex-direction: row;
        justify-content: space-between;
        padding: $m-size;
    }
}

.list-item--message{
    align-items: center;
    justify-content: center;
    color:$dark-gray-2;
    padding:$m-size;
    &:hover{
        background:none;
    }
}

.list-item__title{
    margin: 0;
    word-break: break-all;
}

.list-item__subtitle{
    color:$dark-gray-2;
    font-size: $font-size-small;
    margin:0;
}
.list-item__subdata{
    color: $dark-gray-2;
    word-break: break-all;
    margin:0;
    font-size: $font-size-small;
}

.list-item__data{
    margin: $s-size 0 0 0;
    @media (min-width:$desktop-breakpoint){
        margin: 0;
        padding-left: $s-size;
    }
}