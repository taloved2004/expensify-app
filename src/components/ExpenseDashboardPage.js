import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummay from './ExpensesSummary';
const ExpenseDashboardPage = () => (
    <div>
        <ExpenseSummay />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;