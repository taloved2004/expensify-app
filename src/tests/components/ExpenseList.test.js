import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixures/expenses';

test('should render expense list with givin expenses list', ()=>{
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render a massgae when empty array', () =>{
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
});