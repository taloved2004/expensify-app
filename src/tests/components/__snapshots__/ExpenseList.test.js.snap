// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a massgae when empty array 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item--message"
    >
      <span>
        No expenses
      </span>
    </div>
  </div>
</div>
`;

exports[`should render expense list with givin expenses list 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <Item
      amount={195}
      createdAt={0}
      description="Gum"
      id="1"
      key="1"
      note=""
    />
    <Item
      amount={109500}
      createdAt={-345600000}
      description="Rent"
      id="2"
      key="2"
      note="s"
    />
    <Item
      amount={4500}
      createdAt={345600000}
      description="Credit card"
      id="3"
      key="3"
      note=""
    />
  </div>
</div>
`;
